### PR TITLE
Add test_bar in test_datetime

### PR DESCRIPTION
**Pull request by kots14**
## PR summary

**Version 4:**
Modified tests and raised a revision

**Version 4 Image output -**
TestBar

**Version 3:**
Modified tests and raised a revision

**Version 3 Image output -**
TestBar

**Version 2:**
Modified tests as per [this](https://github.com/matplotlib/matplotlib/pull/27033#pullrequestreview-1665154088) and [this](https://github.com/matplotlib/matplotlib/pull/27033#issuecomment-1753602391)

**Version 2 Image output -**
RevisionBarGraphMatplotlib

**Version 1:**
Adding code for `test_bar` method in `test_datetime.py` mentioned in [#26864](https://github.com/matplotlib/matplotlib/issues/26864)

**Version 1 Image output -**
matplotlib_bar

## PR checklist

- [x] resolves [#26864](https://github.com/matplotlib/matplotlib/issues/26864)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [x] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Review by ksunden — CHANGES_REQUESTED** [review]
I think the bar graphs with dates on the vertical (for `bar`, horizonatal for `barh`) axis should be given a bit more care here. That there is an implicit `0` mark, which results in `1970-01-01` is not a particularly good representation of a bar plot.

I think that for this, there should be a `bottom` argument specified as a datetime and then using a `timedelta` instead of a `datetime` for the heights, as indicated by the `bar` [docstring](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.bar.html#matplotlib.axes.Axes.bar)

Here is some code I wrote for someone who had started working on this but never opened a PR:

```python
import matplotlib.pyplot as plt
import numpy as np
from datetime import datetime, timedelta
import matplotlib as mpl

mpl.rcParams["date.converter"] = 'concise'
fig, (ax1, ax2) = plt.subplots(2, 1, layout='constrained')
price_date = np.array([  datetime(2020, 6, 30),
                datetime(2020, 7, 22),
                datetime(2020, 8, 3),
                datetime(2020, 9, 14)], dtype=np.datetime64)

price_close = [8800, 2600, 8500, 7400]
start_date = np.datetime64(datetime(2020, 6, 1))

ax1.bar(price_date, price_close, width=np.timedelta64(4, "D"))
ax2.bar(np.arange(4), price_date-start_date, bottom=start_date)
```

Note the `start_date` and `price_date-start_date` in this example.

**Comment by ksunden** [comment]
Also, while in your example it is not a bad width, I may suggest also testing the `width` argument (in units of `timedelta`) when the horizontal axis is datetime'd, as that is another place where problems could creep in in this method in particular.

**Comment by kots14** [comment]
Thanks for the suggestions. I'll modify the code and raise a revision.

**Comment by kots14** [comment]
I'll be raising the revision in next 3 days.

**Comment by kots14** [comment]
@ksunden I've raised a revision. Please review the changes.

**Comment by ksunden** [comment]
There still appear to be the implicit 1970 in the new version

And it is very hard to see what is happening in especially the middle two plots (but also the top plot) as the widths overlap all of the bars. A single width is probably sufficient, we just mostly want to test it out that it handles the units as expected, so changing from the default is a good check.

**Comment by kots14** [comment]
Noted.
I'll modify the PR.

**Comment by kots14** [comment]
@ksunden 
I've raised a revision. Please review.

File: lib/matplotlib/tests/test_datetime.py [hunk 1]
```diff
@@ -87,11 +87,26 @@ def test_axvspan(self):
...
+                datetime.datetime(2020, 8, 3),
+                datetime.datetime(2020, 9, 14),
+            ],
+            dtype=np.datetime64,
+        )
+        x_ranges = [8800, 2600, 8500, 7400]
+
+        x = np.datetime64(datetime.datetime(2020, 6, 1))
+        ax1.bar(x_dates, x_ranges, width=np.timedelta64(4, "D"))
+        ax2.bar(np.arange(4), x_dates, bottom=x)
```
Comment (ksunden) [inline lib/matplotlib/tests/test_datetime.py:109]:
```suggestion
        ax2.bar(np.arange(4), x_dates-x, bottom=x)
```

The heights are given in units of _timedelta_, not in position of the top, this is why the plot reaches 2070

**Review by ksunden — CHANGES_REQUESTED** [review]

File: lib/matplotlib/tests/test_datetime.py (hunk 1 above)
Comment (kots14) [reply lib/matplotlib/tests/test_datetime.py]: Done

**Review by ksunden — APPROVED** [review]
bar


I was confused by the image added in the post,which seemed to get rid of the date axis on the y axis of the bottom plot, but I actually do keep it when I run this code, see attached.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**